### DISCUSSION TITLE
Is this project still being developed?

### DISCUSSION THREAD

**Discussion by LucidEye** [Dan-in-CA/SIP#304] [General]
It looks like not much has happened on this project for a while.  Is it still being developed / updated?
I was wondering how well it can communicate with an ESP32 running Tasmota?
Can SIP receive data from external sensors via MQTT? 
For instance, if I have a rain sensor attached to an ESP32 running Tasmota, can SIP use that data to skip a watering cycle if the rain sensor is tripped?
Or, if I have soil moisture sensors attached to an ESP32 running Tasmota, can SIP receive that data via MQTT and be programmed to water only when the soil moisture reaches a determined moisture level?
Can SIP send on/off commands to an ESP32 running Tasmota connected to a relay?
![Screenshot at 2023-06-09 21-39-25](https://github.com/Dan-in-CA/SIP/assets/12551280/f84c19c3-f57c-4e3c-a471-7b31c82d19d1)

I just finished building a simple irrigation valve control box using a 24VAC doorbell transformer, an AC to DC buck converter, a 4-channel relay board, and an ESP32 running Tasmota to control the relay.
![IMG_20230611_203227](https://github.com/Dan-in-CA/SIP/assets/12551280/49aa802b-2521-47af-8b9a-f3a8f0e5bf5c)

The ESP32 receives on/off commands for the relay over WiFI from a Raspberry Pi currently running Mycodo environmental automation software and Mosquitto MQTT Broker. (https://kizniche.github.io/Mycodo/)
![Screenshot at 2023-06-08 14-54-37](https://github.com/Dan-in-CA/SIP/assets/12551280/8d1858de-2939-4020-aa51-b0d7f81e9340)

The ESP32 also has an I2C temp & humidity sensor inside the control box to monitor for any overheating components or possible water intrusion.  This sensor data is transmitted over mqtt to the raspberry pi, where Mycodo can use it to send me a warning message if needed.
![Screenshot at 2023-06-11 20-58-16](https://github.com/Dan-in-CA/SIP/assets/12551280/8520b782-8925-4363-928b-80c5157397cb)

Is SIP capable of these same features?  I am looking for alternative software to experiment with to see if there are any better options available.
Thank you.

**Comment by Dan-in-CA** [comment]
Very cool project!

Yes. SIP is under continued development. The latest version is in the P3-only branch. It runs under Python 3 and includes a new node-red extension that can communicate with remote sensors using MQTT. I have been experimenting with a node-red flow that reads soil moisture sensors connected to an ESP32 over wifi>mqtt and it is working quite well.
The ESP32 was programmed using ESPEasy:
https://espeasy.readthedocs.io/en/latest/
https://td-er.nl/ESPEasy/

There are also some Python based MQTT plugins.

The P3-only branch will be merged into the master branch as soon as the node-red nodes are posted to the node-red library. This still needs some review of the documentation but is usable now. The SIP node-red nodes can be downloaded from
https://github.com/Dan-in-CA/node-red-sip-irrigation-nodes.

![image](https://github.com/Dan-in-CA/SIP/assets/4706639/a48fdd03-55ed-4e58-9dbe-9f2a4873518d)


Dan

**Reply by LucidEye** [reply]
I don't use node-red.  Is it required to make this work?

**Reply by Dan-in-CA** [reply]
No. You can also use the existing MQTT Plugins.
There are other plugins that you may find useful.

Dan

**Reply by LucidEye** [reply]
Sounds interesting and promising... I'll have to free-up one of my Pis and give it a try.

**Reply by jkaidor** [reply]
I'd be interested in the ability to read sensors.  I have a fairly complicated irrigation system, with a lot of moving parts.  There's a 19th century dug well with a pump near the bottom.  The pump runs off a pair of SLA batteries kept charged with 500W worth of solar cells and a 25A MPPT controller.   The pump is controlled by a relay that's turned on by 24VAC from the SIP system.    There is also an inverter and an ultraviolet sterilizer, and a 1-minute timer to turn the inverter and sterilizer on before the pump.  The pump feeds a veritable forest of drip emitters, organized into 8 stations.

   As I said, a lot of moving parts.   Stuff is always going wrong.  Gopher nibbles on a hose... Drippers clog.  The pump wears out.  The batteries wear out.    A station valve gets stuck open.  Station valve gets stuck closed.   Station valve shorts out electrically, frying the driver on the Pi.  Etc.

   It would be VERY nice to monitor a pressure sensor on the main line to verify that the pump is indeed pumping when it should be pumping.

**Reply by LucidEye** [reply]
@jkaidor It sounds like your irrigation system is spread out over a wide area. You would probably need to use distributed ESP32-based sensors that can monitor your critical system points, and then send that data via WiFi or LoRa back to a central monitoring system that can be programmed to send you a text or email notification if something fails.

**Reply by jkaidor** [reply]
The yard is big.  Almost an acre.  Wifi is spotty out there.    At one point, I had a Ubiquiti Bullet outside, and that worked pretty well...until it stopped working.  This morning, I was out at the Pi installing a USB high-sensitivity wifi interface with an external antenna.  Still needs help, the signal is weak.  Later today, I might go out there and fabricate a reflector for that antenna out of aluminum foil & cardboard.
For these sensors, I prefer wires.

I do have a box full of ESP32 "nodemcu" cards that I was programming on a while back.
I'm a retired embedded software engineer.

**Reply by LucidEye** [reply]
Well, unless you want to run signal wires all over your yard for sensors (which won't work due to i2c wire length limits, signal attenuation, and EMI problems) then your only other option besides WiFi is using LoRa hardware for your remote sensors.

**Reply by jkaidor** [reply]
Back to work.... Starting up the system for the year.  Not gonna rain anymore.  The well pump came right up.  Big leak at the connection to the uphill manifold.  A gopher-nibbled hose gushing a few inches under the ground.  

Signal wires would work, but not with I2C of course.  Slow CAN-bus would work.    But yeah, that's a lot of wire.  I just got a wifi 
extender for the back yard.  I'm going to put it inside a PVC outdoor box and bolt it to the back of the sauna.  There's AC there.  Or - I might put it inside the spa.  Plenty of AC in there too.

I just got a couple of Raspberry Pi Pico 2 W's.  Those can connect via wifi.  Their most endearing quality?   They support SWD aka "Serial Wire Debug".   Breakpoints.  Source-level single step.

**Comment by Dan-in-CA** [comment]
Actually this sounds like a case for Modbus over RS485. The ware length can be up to 4000 feet (1200 meters).
You can use an RS485 adapter that plugs into a USB port on the raspi and not even use any GPIO pins. Even though the Modbus spec calls for twisted pair I have used regular burial type sprinkler wire with no problems at up to 100 feet.

Dan

**Reply by jkaidor** [reply]
Modbus.... I think the solar controller supports that.  It's a $$$ 25A Morningstar.

**Comment by jkaidor** [comment]
On 2023-10-07 15:47, Dan wrote:
> Actually this sounds like a case for Modbus over RS485. The ware
> length can be up to 4000 feet (1200 meters).

*** Does it support daisy chaining?   Twisted pair, no problem.  Outdoor 
rated Ethernet cable is cheap.  It doesn't need to be fast.  I'd
like to have multiple pressure sensors and at least one flow sensor.

                      - Jerry

**Reply by LucidEye** [reply]
Modbus is for indoors, factories and manufacturing automation applications.  LoRa is for outdoors, and distances up to 5 miles... which is why it is becoming somewhat of the new "de-facto standard" in agricultural IoT applications like remote sensor and data-logger monitoring. Modbus would require that you snake one _**really long**_ cable all over the property so that you can connect all of your daisy-chained sensors or control modules (like relays) to that one cable.... running that much wire all over the property isn't really practical. I'd also seriously be afraid to have a lot of long wires like that spread over a wide area, and all of them leading back to my expensive automation system, especially if it were housed in my living space... how do you plan on arresting potential lightning strikes on that big copper conductor you just laid out all over your yard?  Here's some more reasons you might not want to use modbus for outdoor agricultural applications.   
https://www.rtautomation.com/rtas-blog/when-not-to-use-modbus/#:~:text=Don't%20use%20Modbus%20to%20represent%20any%20kind%20of%20complicated,in%20a%20Modbus%20register%20table.

**Reply by LucidEye** [reply]
Modbus is 50 years old.  While it still sees a lot of use, it really is ageing out and being replaced with newer and much more capable IoT protocols.

https://www.rtautomation.com/rtas-blog/what-modbus-doesnt-do-very-well/

**Reply by DarthKegRaider** [reply]
I would be surprised if all sensible business adopted wireless.  We don't use any wireless modbus device, all wired.  If something doesn't start or stop when it should, the potential to kill someone exists.  Timings to the ms exist in our operations for a reason, and wireless doesn't offer the best stability for this to happen.

IoT is also not safe in a control systems / manufacturing / production environment.  Keeping critical systems air gapped and off internet should be at the forefront of any CSE.  This way it can't be compromised by some random crypto thief, and your business will keep going.

**Reply by LucidEye** [reply]
Um... WTF are you even talking about??? 
We're talking about an irrigation system, not an enterprise network for a Fortune 500 corporation.  
And yeah, I fully understand the need for hard-wired inputs and outputs in an industrial control environment, but we also aren't talking about an industrial environment here, we're talking about the irrigation system in the back yard.
The only thing that might get "killed" if the sprinklers don't turn on once just because the wifi was on the fritz that day... is a tomato plant. 
And wireless IoT is just as safe and secure as wired hardware.  
Ever heard of data encryption or passwords or secure keys?  
Do you really think that a determined attacker wouldn't be able to cut into a hard-wire data line and infiltrate a data network on the physical layer?  
Have you ever heard of SCADA?  You know, the monitoring systems that most municipalities use to keep an eye on pretty much all critical infrastructure... a lot of SCADA today relies on... wireless technologies!
So yeah... the next time a "random crypto thief" tries to hijack the ESP32 running my sprinkler system and holding it for ransom... I'll think of you and LAUGH MY ASS OFF!!!!!!
P.S.   You should change your handle to "Dunning–Kruger".

**Reply by DarthKegRaider** [reply]
Sure bud, you do you, although snide name calling isn't necessary buddy.  Just makes you rude.

What I was referring to was your comment about modbus being replaced by IoT protocols.  That won't happen for decades in the already established operations.  THIS is a fantastic project running on my Pi and watering my yard and tomato plant using IoT, via secure wifi on it's own VLAN.  What you suggested about modbus "being replaced" with IoT is quite frankly, decades away.  

I'll continue using ladder logic old on my dozens of SCADA plcs for the long foreseeable future in my workplace.  I'll let you go back to attacking some other random person on the internet.

**Comment by Dan-in-CA** [comment]
It supports up to 32 slave units.

**Comment by jkaidor** [comment]
On 2023-10-07 19:10, LucidEye wrote:
> LoRa is for outdoors, and distances up to 5 miles...

*** Can it do peer-to-peer?  Or does there have to be a "gateway"?  I 
note that those are more expensive.

                    - Jerry

**Reply by LucidEye** [reply]
Yes, the hardware would be more expensive for sure.  But you can run them off a small LiPo battery and a mini solar panel and put them anywhere.  The ESP32 does have a WiFI peer-to-peer mode called ESP-NOW.  If you use the peer-to-peer mode for the ESP32s, you would need a gateway in between the peer-to-peer and your LAN networks. You could use one of these for the gateway....
https://www.amazon.com/JacobsParts-WT32-ETH01-Ethernet-Bluetooth-Development/dp/B08WC15NWZ/ref=sr_1_6?crid=297YT6V3OS81G&keywords=ESP32+with+ethernet&qid=1696786903&sprefix=esp32+with+etherne%2Caps%2C145&sr=8-6

If you are set on running wires, you could just run outdoor CAT-5e from a router out to each of these... you could also power them using PoE.

The LoRa modules are more expensive, but no worries about wires...
https://www.amazon.com/MakerFocus-Integrated-Bluetooth-Development-Intelligent/dp/B0BRKRG149/ref=sr_1_7?crid=9RJ97389D9GD&keywords=ESP32+with+LoRa&qid=1696787323&sprefix=esp32+with+lora%2Caps%2C146&sr=8-7

They even have ESP32-LoRa modules with built-in battery holders and charger, you just need to hook up a small solar panel.  And the ESP32 has a deep-sleep mode for maximum power saving when running on battery.
https://www.amazon.com/Wireless-Module-T-Beam-433Mhz-Battery/dp/B07X1NRWXY/ref=sr_1_7?crid=K8OO1P8PAQZQ&keywords=esp32+lora+battery&qid=1696787674&sprefix=ESP32+LoRa+and+battery%2Caps%2C135&sr=8-7&ufe=app_do%3Aamzn1.fos.006c50ae-5d4c-4777-9bc0-4513d670b6bc

**Comment by jkaidor** [comment]
On 2023-10-08 10:56, LucidEye wrote:

> 
> The LoRa modules are more expensive, but no worries about wires...
> https://www.amazon.com/MakerFocus-Integrated-Bluetooth-Development-Intelligent/dp/B0BRKRG149/ref=sr_1_7?crid=9RJ97389D9GD&keywords=ESP32+with+LoRa&qid=1696787323&sprefix=esp32+with+lora%2Caps%2C146&sr=8-7

*** So will these modules do peer-to-peer?  Just buy two of them and 
program them to talk to each other?

                   - Jerry

**Reply by LucidEye** [reply]
Yes, LoRa can do both "star topology" (hub and endpoints), or mesh (peer to peer).  
But you don't just want them to talk to each other, they need to also talk to your LAN otherwise how are you going to access the data?
With a mesh network you would also need a "gateway" device to connect the mesh and your LAN networks.

Also, a cheaper option would be to just use the ESP32's own built-in mesh called ESP-NOW.
The ESP32 mesh network would probably take care of any range issues on an acre sized plot.  
We have over half an acre and I have no problem getting an ESP32 dev kit board to connect to the router inside the house with at least a 50-60% signal strength.
So with a mesh network that range would be significantly increased if you had a few nodes placed fairly evenly around the property.

Supposedly Tasmota now supports ESP32 mesh, but I haven't had time to try it myself yet.
I think ESPHome also has an ESP-NOW module.

But you'll still need some kind of a "go-between" or "gateway" device that can interface with both the mesh and your WiFi or Ethernet LAN if you're trying to get MQTT data from the mesh sensors to SIP or some other automation software that lives on the LAN side of your network.
https://emariete.com/en/gateway-espnow-mqtt/

I bet if you did some digging and asking around on the forums you would find someone who has already written code that allows you to use a single ESP32 that connects to the ESP-NOW mesh with it's on-board antenna, and then connect directly to the Raspberry Pi via I2C.

**Comment by jonathan-marsh** [comment]
Hi Jerry,

 

I’ve also wanted such a feature in the past, though I’ve gotten my system reliable enough through incremental improvements over the years that pressure alerts are no longer as critical.  I went as far as installing a pressure sensor, driven by Arduino and hooked up to my RPi so I could read pressure remotely, but not all the way to hooking up alerts or integrating with SIP.

 

Are you thinking the sensors would be read directly by SIP, or driven a separate process?

 

-Jonathan

**Reply by Kali-** [reply]
I'm been using SIP with Tasmota (SONOF 4CH R3) for almost 2y with "Cli Control" plugin with mosquitto_pub cmq and it work very well, but this topic is interesting :)
